File: nim/search/linear_search.nim
import strformat

func loopSearch(arr: openArray[int], key: int): int =
  for i in 0 ..< arr.len():
    if arr[i] == key: return i

  return -1

func recursiveSearch(
  arr: openArray[int], 
  key: int, index: int
): int =
  if index <= 0: return -1
  let target = index - 1
  if arr[target] == key: return target
  return recursiveSearch(arr, key, target)

proc main() =
  let key = 80
  let arr = [10, 50, 30, 70, 80, 60, 20, 90, 40]
  # let i = loopSearch(arr, key)
  let i = recursiveSearch(arr, key, arr.len())

  if i == -1: echo "Key not found"
  else: echo &"Key Found at index: {i}"

when isMainModule:
  main()

File: clang/search/linear_search.c
#include <stdio.h>

int loop_search(int *arr, int n, int key) {
  for (int i = 0; i < n; i++) {
    if (arr[i] == key) {
      return i;
    }
  }
  return -1;
}

int recursive_search(int *arr, int n, int key) {
  if (n == 0) {
    return -1;
  }

  if (arr[n - 1] == key) {
    return n - 1;
  }

  return recursive_search(arr, n - 1, key);
}

int main() {
  int key = 30;
  int arr[] = {10, 50, 30, 70, 80, 60, 20, 90, 40};
  int length = sizeof(arr) / sizeof(arr[0]);
  // int i = loop_search(arr, length, key);
  int i = recursive_search(arr, length, key);

  if (i == -1)
    printf("Key Not Found");
  else
    printf("Key Found at Index: %d", i);

  return 0;
}
